--- libavformat/matroskaenc.c	2018-08-10 22:18:59.051659000 +0200
+++ libavformat/matroskaenc2.c	2018-08-10 22:18:31.932030400 +0200
@@ -2671,8 +2671,8 @@
     }
 
     for (i = 0; i < s->nb_streams; i++) {
-        // ms precision is the de-facto standard timescale for mkv files
-        avpriv_set_pts_info(s->streams[i], 64, 1, 1000);
+        // ms precision is the de-facto standard timescale for mkv files (Patched: We're using microseconds!!)
+        avpriv_set_pts_info(s->streams[i], 64, 1, 1000000);
     }
 
     return 0;
